fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMultiTargetByRecords.java

```diff
@@ -165,8 +198,7 @@ private CompletableFuture<Boolean> buildRangeOnly(@Nonnull FDBRecordStore store,
                                           lastResult.get().get().getPrimaryKey() :
                                           rangeEnd)
                     .thenCompose(cont -> insertRanges(store.ensureContextActive(), targetRangeSets, packOrNull(rangeStart), packOrNull(cont))
-                                .thenApply(ignore -> !Objects.equals(cont, rangeEnd)));
-
+                                .thenApply(ignore -> !allRangesExhausted(cont, rangeEnd)));
```

Comment:
This also feels like maybe it works in practice because the wrong kind of range is never put into the range set...but that feels a little delicate to me. It's possible I'm missing part of some bigger picture, but it seems like we'd want to retain the ability in general to have a non-contiguous range built so that we can split up the records subspace and build the index in parallel, and a change like this would make that more difficult

Reply:
I was hoping to avoid the use of predefined ranges that are provided by the users in parallel indexing. At the current indexing state, the iteration should stop only when the index is fully built.

Reply:
My problem with that is that this is not the current API of the indexer, and so making these changes seems premature unless you _also_ remove the `startBytes` and `endBytes` from the API (or require that `startBytes` and `endBytes` are both `null` at runtime). Which could certainly be done, and I'm fine with merging these changes after removing the endpoints we don't want to support any more.

I think if we want to understand why the fragmented indexing tests failed with the initial version of this PR but succeeded with the change to `allRangesExhausted`, I think what's going on is that when there is a range in the range set/an end range passed to this method, `allRangesExhausted` will always return `false`, so it will keep looping, until such time as the range set has no more ranges within it, at which point the earlier check (commented "no more missing ranges") fires. This wasn't a problem with the single index indexer because it uses the `hasMore` in a different way from the multi-target case.